cmd/snap-repair/runner.go

```diff
@@ -67,24 +82,145 @@ func (r *Repair) SetStatus(status RepairStatus) {
+	// to run something right after fork() in the child to
+	// know the fd. However because go will close all fds
+	// except the ones in "cmd.ExtraFiles" we are safe to set "3"
+	env = append(env, "SNAP_REPAIR_STATUS_FD=3")
+	env = append(env, "SNAP_REPAIR_RUN_DIR="+rundir)
+
+	workdir := filepath.Join(rundir, "work")
+	if err := os.MkdirAll(workdir, 0700); err != nil {
+		return err
+	}
+
+	cmd := exec.Command(script)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	cmd.Env = env
+	cmd.Dir = workdir
+	cmd.ExtraFiles = []*os.File{statusW}
+	cmd.Stdout = logf
+	cmd.Stderr = logf
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+	statusW.Close()
 
-	// XXX actually run things and captures output etc
+	// wait for repair to finish or timeout
+	killTimerCh := time.After(defaultRepairTimeout)
+	doneCh := make(chan error)
+	go func() {
+		// ignore error, we only care about what we got via
+		// the status pipe
+		doneCh <- cmd.Wait()
+		close(doneCh)
+	}()
+	select {
+	case err = <-doneCh:
+		// done
+	case <-killTimerCh:
+		if err := osutil.KillProcessGroup(cmd); err != nil {
+			logger.Noticef("cannot kill timed out repair %s: %s", r, err)
+		}
+		err = fmt.Errorf("repair did not finish within %s", defaultRepairTimeout)
+	}
+	if err != nil {
+		err = fmt.Errorf("%q failed: %s", r.Ref(), err)
+
+		statusPath := filepath.Join(rundir, baseName+".retry")
+		if err := os.Rename(logPath, statusPath); err != nil {
+			logger.Noticef("cannot write stamp: %s", statusPath)
+		}
+
+		if err := r.errtrackerReport(err, statusPath); err != nil {
+			logger.Noticef("cannot report error to errtracker: %s", err)
+		}
+		// ensure the error is present in the output log
+		fmt.Fprintf(logf, "\n%s", err)
+
+		return err
+	}
+
+	// read repair status pipe, use the last value
+	var status RepairStatus
+	scanner := bufio.NewScanner(statusR)
+	for scanner.Scan() {
+		switch strings.TrimSpace(scanner.Text()) {
+		case "done":
+			status = DoneStatus
+		case "skip":
+			status = SkipStatus
+		}
+	}
+	if scanner.Err() != nil {
+		return err
+	}
+
+	statusPath := filepath.Join(rundir, baseName+"."+status.String())
+	if err := os.Rename(logPath, statusPath); err != nil {
+		return err
+	}
+	r.SetStatus(status)
 
 	return nil
 }
 
+// errtrackerReport reports an repairErr with the given logPath to the
+// snap error tracker.
+func (r *Repair) errtrackerReport(repairErr error, logPath string) error {
+	// FIXME: instead of a pseudo snap we could also just use a
+	// different "ProblemType". Right now we always use
+	// "ProblemType: snap" but for repairs we could set it to
+	// "ProblemType: snap-repair" or similar
+	pseudoSnap := r.Ref().String()
+
+	errMsg := fmt.Sprintf("%s", repairErr)
+
+	scriptOutput, err := ioutil.ReadFile(logPath)
+	if err != nil {
+		logger.Noticef("cannot read %s", logPath)
+	}
+	dupSig := fmt.Sprintf("%s\noutput:\n%s", errMsg, scriptOutput)
+
+	extra := map[string]string{
+		"Revision": strconv.Itoa(r.Revision()),
+	}
```

Comment:
we want to also add here brand-id, repair-id,  and the status  if we change the code to mark thing as done